File: 36-valid-sudoku/36-valid-sudoku.java
import java.util.HashSet;

class Solution {
    public boolean isValidSudoku(char[][] board) {
        //For in element in every array we add the following String to a set (if the element is not ".")
            //'5' in row 0
            //'5' in column 0
            //'5' in cube (0,0)
        //if one of them cannot be added to a set, return false, because it means 5 occured again in the same row/column/cube
        //To calculate cube, we can use the loop
            //horizontal = 3
        
        Set seen = new HashSet();
        for(int i = 0; i < 9; i++){ //verti
            for(int j = 0; j < 9; j++){ //hori
                if(board[i][j] != '.'){
                    if(!seen.add(board[i][j] + " in row " + i) ||
                        !seen.add(board[i][j] + " in column " + j) ||
                            !seen.add(board[i][j] + " in cube " + i/3 + "," + j/3)){
                        return false;
                    }
                }
            }
        }
        return true;
    }
}

File: 36-valid-sudoku/36-valid-sudoku.py
"""
Use a set to save the positions. If we can't add it into our set, then it's repeated, return false.
For each char that is not ".", we add three entries to our set. Examples:
"5 in row 0" -> num in row j
"5 in column 0" -> num in column i
"5 in cube 0,0", 1st 0 = i//3, and 2nd 0 is j//3

Time = O(n), Space = O(n) 

"""
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        resSet = set()
        for i, row in enumerate(board):
            for j, c in enumerate(row):
                if c != ".":
                    row = c + " in row " + str(i)
                    column = c + " in column " + str(j)
                    cube = c + " in cube " + str(i//3) + "," + str(j//3)
                                
                    if row in resSet or column in resSet or cube in resSet:
                        return False
                    else:
                        resSet.update([row, column, cube])
                    
        return True